cpp/src/arrow/compute/registry.cc

```diff
@@ -48,23 +61,56 @@ class FunctionRegistry::FunctionRegistryImpl {
     if (it != name_to_function_.end() && !allow_overwrite) {
       return Status::KeyError("Already have a function registered with name: ", name);
     }
-    name_to_function_[name] = std::move(function);
+    add(name, std::move(function));
```

Comment:
```suggestion
    if (do_add) {
      name_to_function_[name] = std::move(function);
    }
```
If we change `add` to `bool do_add` I think this is a bit easier to follow.  If performance is a concern (I don't think it would be as this isn't really in a critical section) you could use a template...

```
template <bool kDoAdd>
Status DoAddFunction(std::shared_ptr<Function> function, bool allow_overwrite) {
  ...
  if (kDoAdd) {
    name_to_function_[name] = std::move(function);
  }
}
virtual Status CanAddFunction(std::shared_ptr<Function> function,
                              bool allow_overwrite) {
  return DoAddFunction<false>(function, allow_overwrite);
}
...

```